OLED Driver Feature (#5288)
qmk/qmk_firmware: docs/feature_oled_driver.md

# OLED Driver

## OLED Supported Hardware

128x32 OLED modules using SSD1306 driver IC over I2C. Supported on AVR based keyboards. Possible but untested hardware includes ARM based keyboards and other sized OLED modules using SSD1306 over I2C, such as 128x64. 

> **WARNING:** This OLED Driver currently uses the new i2c_master driver from split common code. If your split keyboard uses i2c to communication between sides this driver will most likely cause a conflict (serial is fine). Please contact your keyboard vendor and ask them to migrate to the latest split common code to fix this. Message @mechmerlin#2999 in the qmk discord for more details. (ps: he likes rick-rolls)

Reviewer: In theory, as long as the i2c address doesn't overlap.... it shouldn't be an issue.  But I suspect that you'd have to add some handling for this.  

Also, there was this: https://github.com/mtei/qmk_firmware/issues/7

Author: Ya, it was more a warning that I haven't tested it and many keyboards are using the old i2c source as well. That and a slight nudge to get more keyboards moved to split common. Interesting link, saved it to read later.

Reviewer: I think a less dire warning would be better.

Also, `!>` will create a warning note. https://docs.qmk.fm/#/documentation_best_practices?id=important

Reviewer: ```suggestion
!> Warning: This OLED Driver currently uses the new i2c_master driver from split common code. If your split keyboard uses i2c to communication between sides, this driver will most likely cause an address conflict (serial is fine). Please contact your keyboard vendor and ask them to migrate to the latest split common code to fix this. Message @mechmerlin#2999 in the qmk discord for more details. (ps: he likes rick-rolls)
```

Author: Fixed

|   |   |   |   |   |   |
|---|---|---|---|---|---|
| 0 | 1 |   |   |   |   |
| 2 | 3 |   |   |   |   |
| 4 | 5 |   |   |   |   |
| 6 | 7 |   |   |   |   |

However the local buffer is stored as if it was Height x Width display instead of Width x Height, e.g:

|   |   |   |   |   |   |
|---|---|---|---|---|---|
| 3 | 7 |   |   |   |   |
| 2 | 6 |   |   |   |   |
| 1 | 5 |   |   |   |   |
| 0 | 4 |   |   |   |   |

So those precalculated arrays just index the memory offsets in the order in which each one iterates its data.

## OLED API

```C++
// Initialize the OLED display, rotating the rendered output 180 degrees if true.
// Returns true if the OLED was initialized successfully

> bool oled_init(bool flip180);

Reviewer: ```suggestion
bool oled_init(uint8_t rotation);
```

Author: fixed

// Called at the start of oled_init, weak function overridable by the user
// flip180 - the value passed into oled_init
// Return true if you want the oled to be flip180

> bool oled_init_user(bool flip180);

Reviewer: ```suggestion
uint8_t oled_init_user(uint8_t rotation) {
```

Author: fixed

# OLED Driver

## OLED Supported Hardware

128x32 OLED modules using SSD1306 driver IC over I2C. Supported on AVR based keyboards. Possible but untested hardware includes ARM based keyboards and other sized OLED modules using SSD1306 over I2C, such as 128x64. 

!> Warning: This OLED Driver currently uses the new i2c_master driver from split common code. If your split keyboard uses i2c to communication between sides this driver could cause an address conflict (serial is fine). Please contact your keyboard vendor and ask them to migrate to the latest split common code to fix this. 

## Usage

To enable the OLED feature, there are three steps. First, when compiling your keyboard, you'll need to set `OLED_DRIVER_ENABLE=yes` in `rules.mk`, e.g.:

```
BOOTMAGIC_ENABLE = no
MOUSEKEY_ENABLE = no
STENO_ENABLE = no

> EXTRAKEY_ENABLE = yes

Reviewer 2: Can you remove the extraneous lines? We've found it's easier to understand when you limit it to only the most relevant snippets.

Author: fixed

OLED_DRIVER_ENABLE = yes
```

This enables the feature and the `OLED_DRIVER_ENABLE` define. Then in your `keymap.c` file, you will need to implement the user task call, e.g:

```C++
#ifdef OLED_DRIVER_ENABLE
void oled_task_user(void) {
  // Host Keyboard Layer Status
  oled_write_P(PSTR("Layer: "), false);
  switch (biton32(layer_state)) {
    case _QWERTY:
      oled_write_P(PSTR("Default\n"), false);
      break;
    case _FN:
      oled_write_P(PSTR("FN\n"), false);
      break;
    case _ADJ:
      oled_write_P(PSTR("ADJ\n"), false);
      break;
    default:

>       // Or use the write_ln shortcut

Reviewer 2: I'm not sure what this comment means

Author: fixed

      oled_write_P(PSTR("Undefined\n"), false);
  }

  // Host Keyboard LED Status
  uint8_t led_usb_state = host_keyboard_leds();
  oled_write_P(led_usb_state & (1<<USB_LED_NUM_LOCK) ? PSTR("NUMLCK ") : PSTR("       "), false);
  oled_write_P(led_usb_state & (1<<USB_LED_CAPS_LOCK) ? PSTR("CAPLCK ") : PSTR("       "), false);
  oled_write_P(led_usb_state & (1<<USB_LED_SCROLL_LOCK) ? PSTR("SCRLCK ") : PSTR("       "), false);
}
#endif
```

## Other Examples

In split keyboards, it is very common to have two OLED displays that each render different content and oriented flipped differently. You can do this by switching which content to render by using the return from `is_keyboard_master()` or `is_keyboard_left()` found in `split_util.h`, e.g:

```C++
#ifdef OLED_DRIVER_ENABLE
oled_rotation_t oled_init_user(oled_rotation_t rotation) {
  if (!is_keyboard_master())
    return OLED_ROTATION_180;  // flips the display 180 degrees if offhand
  return rotation;
}

void oled_task_user(void) {
  if (is_keyboard_master()) {
    render_status();     // Renders the current keyboard state (layer, lock, caps, scroll, etc)
  } else {

>     render_logo();       // Renders a statuc logo

Reviewer 2: What is render_logo? Seems like that's only in keymaps but not mentioned in the docs?

Author: fixed
